### PR TITLE
Gravatar OAuth in SwiftUI

### PR DESCRIPTION
**Pull request by etoledom**
### Description
This PR implements the Gravatar OAuth flow.

CleanShot 2024-08-22 at 14 51 23

### Testing Steps

- Build and run the SwiftUI demo app
- Go to Profile Editor with oauth demo screen.
- insert an email and open the editor.
  - Check that OAuth flow works as expected.

### CODE REVIEW

File: Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+enum OAuthError: Error {
+    case notConfigured
+    case couldNotCreateOAuthURLWithGivenSecrets
+    case couldNotGetCodeFromURL(String)
+    case oauthResponseError(String)
+    case unknown(Error)
+    case couldNotStoreToken(Error)
+}
```
Comment (etoledom) [inline Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift]: This became a bit messy and I don't like it:

- `notConfigured`: Secrets are not set up. This maybe could be a `assertionFailure` instead of throwing? The idea is to warn the developer.
- `couldNotCreateOAuthURLWithGivenSecrets` this comes from using `Encodable` to add the query items to the URL. Probably if we add them manually we wouldn't need this error. And if this error happens, it's probably a developer mistake which should be catch on dev time (maybe assertion also?)
- `case couldNotGetCodeFromURL(String)` error on parsing the callback URL to get the code. This could be a runtime error but not sure how could this be handled 🤔 .
- `oauthResponseError(String)` Something went wrong at the backend. Needs to be handled in the UI.
- `unknown` ¯\_(ツ)_/¯  ... Probably on the UI also.
- `couldNotStoreToken(Error)` Keychain error. I got errors here because the keychan was wrongly set up. But it was a dev issue. Probably there are instances that this could be a user issue, but again, not sure how this should be handled. Technically we have the token and the user should be able to edit the profile, so I don't think that throwing is a good option. The problem is that they will need to be authenticated again next time.

File: Sources/GravatarUI/SwiftUI/OAuthSession/Keychain.swift [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import Foundation
+
+protocol SecureStorage: Sendable {
+    func setPassword(_ password: String, for key: String) throws
+    func deletePassword(with key: String) throws
+    func password(with key: String) throws -> String?
```
Comment (pinarol) [inline Sources/GravatarUI/SwiftUI/OAuthSession/Keychain.swift]: `password` is not quite the right terminology I think, how about `token` or `secret`? `password` is more for "something I know" kind of secrets. 🤔

File: Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        decoder.keyDecodingStrategy = .convertFromSnakeCase
+        return decoder
+    }()
+
+    init(authenticationSession: AuthenticationSession = OldAuthenticationSession(), storage: SecureStorage = Keychain()) {
+        self.authenticationSession = authenticationSession
+        self.storage = storage
+    }
+
+    func isSession(with email: Email) -> Bool {
```
Comment (pinarol) [inline Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift]: how about `hasSession(with: ...)`?

File: Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        request.httpBody = try params.queryItems.string?.data(using: .utf8)
+
+        return request
+    }
+}
+
+enum OAuthError: Error {
+    case notConfigured
+    case couldNotCreateOAuthURLWithGivenSecrets
+    case couldNotGetCodeFromURL(String)
```
Comment (pinarol) [inline Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift]: this was a bit ambiguous to me so I suggest:

> couldNotParseAccessCode

File: Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift (hunk 1 above)
Comment (pinarol) [reply Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift]: I would add the assertionFailures for `notConfigured` and `couldNotCreateOAuthURLWithGivenSecrets` but not necessarily replace the "throw" statements. We still need the throws for prod. Throws also allow us return non-optional values.

File: Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+    func deleteSession(with email: Email) {
+        try? storage.deletePassword(with: email.rawValue)
+    }
+
+    func sessionToken(with email: Email) -> String? {
+        try? storage.password(with: email.rawValue)
+    }
+
+    func authenticate(with email: Email) async throws -> String {
```
Comment (pinarol) [inline Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift]: It is not easily understandable what this method does and returns just by looking at the signature. `authenticate` is just half of OAuth. The work done here is I think described better by something like:

> acquireAccessToken(for email: Email) async throws -> String

File: Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift (hunk 1 above)
Comment (pinarol) [reply Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift]: > couldNotStoreToken(Error) Keychain error. I got errors here because the keychan was wrongly set up. But it was a dev issue. Probably there are instances that this could be a user issue, but again, not sure how this should be handled. Technically we have the token and the user should be able to edit the profile, so I don't think that throwing is a good option. The problem is that they will need to be authenticated again next time

As you said, since the user is able to continue I think we don't need to re-throw this one. This can be something we swallow inside a separate `do { ... } catch {...}` and handle just with an assertionFailure.

File: Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            return error
+        case let error as NSError:
+            if error.domain == ASWebAuthenticationSessionErrorDomain {
+                return .oauthResponseError(error.localizedDescription)
+            }
+            return .unknown(error)
+        case let error as Keychain.KeychainError:
+            return .couldNotStoreToken(error)
+        default:
+            print("🛑 Error: \(type(of: error)) - \(error.localizedDescription)")
```
Comment (pinarol) [inline Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift]: let's remove the "print" call. We can probably use the Apple’s os.log framework for logging, doesn't have to be in this PR though. 🤔

File: Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    }
+}
+
+enum OAuthError: Error {
+    case notConfigured
+    case couldNotCreateOAuthURLWithGivenSecrets
+    case couldNotGetCodeFromURL(String)
+    case oauthResponseError(String)
+    case unknown(Error)
+    case couldNotStoreToken(Error)
```
Comment (pinarol) [inline Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift:112]: I think `DecodingError` deserves a separate case here.

File: Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+extension OAuthError {
+    static func from(error: Error) -> OAuthError {
+        switch error {
```
Comment (pinarol) [inline Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift:118]: We can catch "DecodingError"s and re-throw them after adding an assertionFailure maybe? This means our decoding keys or types are not matching the response, so the developer should fix it.

```
case let error as DecodingError:
   assertionFailure("Unable to decode the response. Error: \(error.description)")
   throw .decodingError(error)

```

File: Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift [hunk 9]
```diff
@@ -42,7 +42,7 @@ struct AvatarPickerView: View {
                         avatarsLoadingView()
                     }
                 }
-                .task {
+                .onAppear() {
```
Comment (pinarol) [inline Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift]: just curious, why is this change necessary?

File: Sources/GravatarUI/SwiftUI/ProfileEditor/ProfileEditor.swift [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    var body: some View {
+        VStack {
+            if isSession, let token = oauthSession.sessionToken(with: email) {
+                switch entryPoint {
+                case .avatarPicker:
+                    AvatarPickerView(model: .init(email: email, authToken: token))
+                }
+                Button("Log out (for testing only)") {
```
Comment (pinarol) [inline Sources/GravatarUI/SwiftUI/ProfileEditor/ProfileEditor.swift]: This is meant to be a temp Button I guess but I feel like it can be useful during development in the long run as well. So how about we wrap it with some kind of compiler flag and keep it?

```
#if DEMO

#endif
```

**Review by pinarol — COMMENTED** [review]
Looks pretty good overall. Just dropping some non-critical comments and a question.

We need UI for this 2 case but it doesn't have to be in this PR.

- we don’t have the token yet but we’ll try (we need an activity indicating UI)
- we tried to get the token but we failed (we can reuse the ContentLoadingErrorView here as well)

File: Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift (hunk 9 above)
Comment (etoledom) [reply Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift]: Oh, I had a problem where the picker wasn't loading the first time. I thought this could be the problem. It wasn't and I forgot to change it back.

Also, `model.refresh()` is not async, and it opens a task itself, so I think `.task` here is not really necessary 🤔 

Though both options will work just fine.

File: Sources/GravatarUI/SwiftUI/OAuthSession/Keychain.swift (hunk 2 above)
Comment (etoledom) [reply Sources/GravatarUI/SwiftUI/OAuthSession/Keychain.swift]: I used password since is what we set in the keychain, though we use it to store an auth token.

On the implementation we use `kSecClass: kSecClassGenericPassword` (on line 22 for example).

Now, since it's for internal usage, I guess we can change to `secret` which I think is actually a more generic word than `password` ¯\_(ツ)_/¯

File: Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift (hunk 5 above)
Comment (etoledom) [reply Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift]: Sounds good 👍

File: Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift (hunk 6 above)
Comment (etoledom) [reply Sources/GravatarUI/SwiftUI/OAuthSession/OAuthSession.swift]: Sure, this was for debugging and see the types of the error thrown to be catch 👍 
`os.log` is a good idea. It might help developers debug issues on their users.

File: Sources/GravatarUI/SwiftUI/ProfileEditor/ProfileEditor.swift (hunk 10 above)
Comment (etoledom) [reply Sources/GravatarUI/SwiftUI/ProfileEditor/ProfileEditor.swift]: Sounds good!

File: Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift (hunk 9 above)
Comment (pinarol) [reply Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift]: Maybe we should make model.refresh() async. 🤔 

What's good about `.task` is, it will automatically cancel the task when the view is destroyed. `.onAppear()` does not have this feature. I think `.task` is slightly a better practice but we'll need to make model.refresh() async so the cancellation is propagated to child tasks I think 🤔. No need to dive into this in this PR though.

Comment (pinarol) [reply Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift]: I created an issue to investigate this https://github.com/Automattic/Gravatar-SDK-iOS/issues/362

Comment (etoledom) [reply Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift]: > What's good about .task is, it will automatically cancel the task when the view is destroyed.

Makes sense 🤔 

> I created an issue to investigate this https://github.com/Automattic/Gravatar-SDK-iOS/issues/362

Nice! Thanks 🙏

File: Sources/GravatarUI/SwiftUI/ProfileEditor/ProfileEditor.swift (hunk 10 above)
Comment (etoledom) [reply Sources/GravatarUI/SwiftUI/ProfileEditor/ProfileEditor.swift]: @pinarol - I tried this and the Build Conditions from the demo app won't be passed to the package, so there's no access to it from within the SDK.

I decided to move this button to the demo screen, previous to present the picker.

Now we will have a `Log out ` button when we detect a session:

CleanShot 2024-08-22 at 14 28 18@2x

Comment (pinarol) [reply Sources/GravatarUI/SwiftUI/ProfileEditor/ProfileEditor.swift]: I see. That works too. 👍

File: Sources/GravatarUI/SwiftUI/ProfileEditor/ProfileEditor.swift [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+            performAuthentication()
+        }
+    }
+
+    @MainActor
+    func performAuthentication() {
+        Task {
+            isAuthenticating = true
+            if !oauthSession.hasSession(with: email) {
+                _ = try? await oauthSession.retrieveAccessToken(with: email)
```
Comment (etoledom) [inline Sources/GravatarUI/SwiftUI/ProfileEditor/ProfileEditor.swift:50]: The possible errors are still not yet properly handled here, but we can do this alongside building the error screens to present these errors.

**Comment by pinarol** [comment]
Something caught my attention, I see the diff for my secrets file:

Screenshot 2024-08-22 at 15 59 07

Seems like the path doesn't match the paths in the `.gitignore`

**Comment by etoledom** [comment]
@pinarol - I believe this is ready for another look 👀

**Comment by etoledom** [comment]
> Something caught my attention, I see the diff for my secrets file:
> 
> Screenshot 2024-08-22 at 15 59 07
> Seems like the path doesn't match the paths in the `.gitignore`

You can delete this file.

This was an error when we renamed the demo apps. We left this file which is not supposed to be in source control inside the old folder name. Now it shows because I removed its entry from .gitignore.

You get this since this file still exist in your machine, but it's safe to be deleted 👍

File: Demo/Demo/Gravatar-SwiftUI-Demo/DemoProfileEditorViewView.swift [hunk 12]
```diff
@@ -0,0 +1,60 @@
+import SwiftUI
+import GravatarUI
```
Comment (pinarol) [inline Demo/Demo/Gravatar-SwiftUI-Demo/DemoProfileEditorViewView.swift]: typo in the filename ? "DemoProfileEditorViewView.swift"

File: Sources/GravatarUI/SwiftUI/OAuthSession/Keychain.swift [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+protocol SecureStorage: Sendable {
+    func setSecret(_ secret: String, for key: String) throws
+    func deleteSecret(with key: String) throws
+    func secret(with key: String) throws -> String?
+}
+
+struct Keychain: SecureStorage {
+    enum KeychainError: Error {
+        case unexpectedPasswordData
+        case cannotConvertPasswordIntoData
```
Comment (pinarol) [inline Sources/GravatarUI/SwiftUI/OAuthSession/Keychain.swift]: For consistency, let's replace `password` with `secret` here as well.

File: Sources/GravatarUI/SwiftUI/View+Additions.swift [hunk 14]
```diff
@@ -25,3 +25,16 @@ extension View {
             )
     }
 }
+
+@MainActor
+extension View {
```
Comment (pinarol) [inline Sources/GravatarUI/SwiftUI/View+Additions.swift]: no need to add a new `extension View` I think? we already have one in the same file.

**Review by pinarol — APPROVED** [review]
Looks good, just leaving some nits. I don't need to check again.

**Comment by pinarol** [comment]
> You get this since this file still exist in your machine, but it's safe to be deleted 👍

Oh I see. Cool.
